### ISSUE TITLE
Difference in number of oneDNN Matmul operation calls in x86/64 and aarch64

### ISSUE THREAD

**Issue by abhishek-rn** [oneapi-src/oneDNN#1713]
Details of systems used:

ARM System : Graviton 3 (c7g.8xlarge)
Architecture: aarch64
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: ARM
Model: 1
Thread(s) per core: 1
Core(s) per socket: 32
Caches (sum of all):
L1d: 2 MiB (32 instances)
L1i: 2 MiB (32 instances)
L2: 32 MiB (32 instances)
L3: 32 MiB (1 instance)

Intel System: Icelake (c6i.8xlarge)
Architecture: x86_64
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) Platinum 8375C CPU @ 2.90GHz
CPU family: 6
Model: 106
Thread(s) per core: 2
Core(s) per socket: 16
Socket(s): 1
Stepping: 6
Virtualization features:
Hypervisor vendor: KVM
Virtualization type: full
Caches (sum of all):
L1d: 768 KiB (16 instances)
L1i: 512 KiB (16 instances)
L2: 20 MiB (16 instances)
L3: 54 MiB (1 instance)

We ran Tensorflow Hugging Face BERT model for Inference (Python Code Attached as txt file here) with oneDNN verbose and BF16 flags enabled.
[TF_bert_inf.txt](https://github.com/oneapi-src/oneDNN/files/12510063/TF_bert_inf.txt)

The code is run on 2 cores for both the Intel and ARM systems.

Below are the count for matmul operation for x64 and aarch64 for different sizes from oneDNN verbose logs:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Sizes (M:N:K) | Aarch64_Count | x64_Count
-- | -- | --
1x12x12x12:1x12x12x64:1x12x12x64 | 9 | 14
1x12x12x64:1x12x64x12:1x12x12x12 | 11 | 14
1x12x2x2:1x12x2x64:1x12x2x64 | 24 | 10
1x12x2x64:1x12x64x2:1x12x2x2 | 31 | 9

</body>

</html>
 Here, MxNxK is the matmul operation.

Why is there a significant difference in the number of calls for each type of Matmul operation between aarch64 and x86

Thanks

**Comment by abhishek-rn** [comment]
Tensorflow version used is 2.12
Attaching the oneDNN verbose files:
[aarch64_Bert_oneDNN_Verbose.txt](https://github.com/oneapi-src/oneDNN/files/12511014/aarch64_Bert_oneDNN_Verbose.txt)
[x64_Bert_oneDNN_Verbose.txt](https://github.com/oneapi-src/oneDNN/files/12511025/x64_Bert_oneDNN_Verbose.txt)

**Comment by nSircombe** [comment]
Hi @abhishek-rn,

I was just wondering, when adding up matmul calls, are you including calls to `gemm_api` (see https://github.com/oneapi-src/oneDNN/issues/1442)? In the AArch64 build of TensorFlow, these calls will be dispatched to oneDNN's `matmul` interface, rather than the `gemm_api` entry point, as a result of https://github.com/tensorflow/tensorflow/pull/60355. Could this account for some of the `matmul` calls present in the AArch64 case, but not x86?

Also, with this TensorFlow change - https://github.com/tensorflow/tensorflow/pull/61110 - heuristics (active on Neoverse-V1, but not on x86) may choose to dispatch some smaller matmul ops to Eigen, rather than oneDNN, leading to some calls appearing in the x86 log, but not the AArch64 one.

**Comment by abhishek-rn** [comment]
Hi @nSircombe ,

Appreciate your quick response.

As per the oneDNN verbose (previously attached txt file), there are no gemm_api calls for aarch64.
However there are some gemm_api calls for x64. But these calls are for different shapes/sizes and there are corresponding calls for these in aarch64 as matmul. (Refer below):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Sizes | aarch_Count (Matmul) | x64_Count (GEMM_API)
-- | -- | --
12x3072:3072x768 | 12 | 11
12x768:768x3072 | 12 | 12
12x768:768x768 | 48 | 48
1x768:768x768 | 3 | 2
2x3072:3072x768 | 23 | 12
2x768:768x3072 | 23 | 12
2x768:768x768 | 96 | 48



</body>

</html>

We took TF Eager logs and filtered overall operations being called. We observe Graviton(aarch64) having more number of calls compared to Icelake(x64).
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Operations | aarch64 | x64
-- | -- | --
_MklMatMul | 219 | 146
_MklBatchMatMulV2 | 72 | 48
_EagerConst | 1869 | 1563
VarHandleOp | 202 | 202
AssignVariableOp | 401 | 401
Fill | 129 | 126
ZerosLike | 2 | 1
Shape | 45 | 30
StatelessRandomGetKeyCounter | 76 | 76
StatelessTruncatedNormalV2 | 76 | 76
Mul | 376 | 276
AddV2 | 611 | 499
Max | 3 | 2
Less | 3 | 2
All | 3 | 2
ResourceGather | 9 | 6
Identity | 120 | 80
Range | 3 | 2
ExpandDims | 3 | 2
Mean | 150 | 100
StopGradient | 75 | 50
SquaredDifference | 75 | 50
Rsqrt | 75 | 50
ReadVariableOp | 588 | 392
Sub | 78 | 52
Reshape | 579 | 386
Cast | 3 | 2
BiasAdd | 219 | 146
Transpose | 217 | 169
RealDiv | 72 | 48
Softmax | 36 | 24
Erf | 36 | 24
StridedSlice | 3 | 2
Tanh | 3 | 2



</body>

</html>


So it doesn't seem to be the case for operators using Eigen flow for Icelake.
Moreover, it woudn't make sense to have some of the matmul operation routed to oneDNN and some routed to eigen for same shapes.

Attaching TF logs here:
[Bert_max_verbose_TF_aarch64_oneDNN.txt](https://github.com/oneapi-src/oneDNN/files/12519284/Bert_max_verbose_TF_aarch64_oneDNN.txt)
[Bert_max_verbose_TF_x86_oneDNN.txt](https://github.com/oneapi-src/oneDNN/files/12519285/Bert_max_verbose_TF_x86_oneDNN.txt)

Thanks

**Comment by nSircombe** [comment]
Hi @abhishek-rn 

> As per the oneDNN verbose (previously attached txt file), there are no gemm_api calls for aarch64.

yep. that's expected - since https://github.com/tensorflow/tensorflow/pull/60355

> Moreover, it woudn't make sense to have some of the matmul operation routed to oneDNN and some routed to eigen for same shapes.

That might depend on how the op has been dispatched by TF. Where TF is calling `matmul` in place of `gemm_api` on AArch64, I don't _think_ this will be exposed to the `matmul` heuristics - so this may see ops being sent to `matmul` that would be carried out by Eigen if the `matmul` heuristics had been applied (there's room for improvement to the heuristics here, perhaps). 

Ultimately, the differences you're seeing between the two runs is, almost certainly, due to choices made in TF on: which backend to send ops to; how `MklMatMulOp` is handled on AArch64; and changes to graph optimisation. You would need to rollback all these changes in TF to do a straight comparison - but in that case you'll likely see oneDNN's `ref` kernels picking up some of the work, which will have very poor performance relative to Eigen, and overall I would expect you to see a performance hit.

**Comment by abhishek-rn** [comment]
Hi @nSircombe 

Thanks again for the clarification.

With a quick check, we were able to confirm that this mismatching count of calls is specific to TF and non-existent in Pytorch or even if we try to run examples from oneDNN src (matmul perf).

Is there a way to figure out and confirm how many calls are going to Eigen or OneDNN for the same model across x64 and aarch64. We are looking only at Matmul calls within BERT model. oneDNN calls are easy to get but Eigen calls remain illusive.

Should we raise a question on TF for this.
Your comments/suggestions would be valuable.

Thanks

**Comment by abhishek-rn** [comment]
Updates on this:
I used a docker version from ARM using tag: [r23.08-tf-2.13.0-onednn-acl].
Now the TF Eager logs are same as what we get for x64 system for same TF version i.e., 2.13.
Even the Matmul calls are equal for both aarch64 and x64 in oneDNN verbose for different shapes.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Matmul | aarch64_Count | x64_Count
-- | -- | --
1x12x2x64:1x12x64x2:1x12x2x2 | 12 | 12
1x12x2x2:1x12x2x64:1x12x2x64 | 12 | 12
1x12x12x64:1x12x64x12:1x12x12x12 | 12 | 12
1x12x12x12:1x12x12x64:1x12x12x64 | 12 | 12
Total | 48 | 48



</body>

</html>

Even where x64 calls gemm_api, the corresponding matmul calls in aarch64 are equal.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Matmul/GemmAPI | aarch64_Count | x64_Count
-- | -- | --
2x768:768x768 | 48 | 48
2x768:768x3072 | 12 | 12
2x3072:3072x768 | 12 | 12
1x768:768x768 | 2 | 2
12x768:768x768 | 48 | 48
12x768:768x3072 | 12 | 12
12x3072:3072x768 | 12 | 12
Total | 146 | 146



</body>

</html>

Now, since reorder operator is part of these matmul and gemm api calls, they ideally should have one to one correspondence.
However, here is where the difference pops up.
The number of reorder calls are different between x64 and aarch64.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Reorder | aarch64_Count | x64_Count
-- | -- | --
1x2x12x64 | 0 | 36
1x12x2x64 | 12 | 12
768x768 | 98 | 49
3072x768 | 24 | 12
768x3072 | 24 | 12
1x12x12x64 | 12 | 48
1x12x64x2 | 12 | 0
1x12x64x12 | 12 | 0
Total | 194 | 169



</body>

</html>

The larger sizes of reorder have exactly twice the number of calls for aarch64 compared to x64. If we try to match the reorder calls with Matmul(aarch64)/Gemm_API(x64) calls for larger sizes, we clearly see that Matmul:Reorder calls are 1:1 in case of aarch64 and in case of x64, we see Gemm_API:Reorder is 1:1/2. 
Is there some kind of caching of operations happening in x64?

Also, another thing to notice is that for smaller matrix sizes, the reorders do not match with matrix calls and there are some extra reorders for sizes in aarch64
The size 1x2x12x64 do not feature in aarch64 and also do not feature in matmul calls for either x64 or aarch64, however it does have about 36 calls in reorder for x64.  Not sure why this is being called. Is it for some intermediate reorder case which we are missing.

Now the difference is at the oneDNN level.
What could be the reasons for this

**Comment by abhishek-rn** [comment]
To make testing easier, we ran the cnn-inference-f32 example in the repo with oneDNN verbose:

The count for all the operators in oneDNN is same except for reorder operator between aarch64 and x64 as seen below:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ABHISH~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Shapes | aarch64 | x64
-- | -- | --
96x3x11x11 | 1 | 1
2x128x48x5x5 | 0 | 1
384x256x3x3 | 1 | 1
2x192x192x3x3 | 0 | 1
2x128x192x3x3 | 0 | 1
4096x4096 | 0 | 1
1000x4096 | 0 | 1
1x3x227x227 | 100 | 100
1x256x6x6 | 0 | 100
1x1000 | 100 | 100
1x96x27x27 | 100 | 0
1x256x13x13 | 100 | 0
1x384x13x13 | 100 | 0
Total | 502 | 307



</body>

</html>

Why would there be a separate set of reorder initiated for the same model between aarch64 and x64 

The oneDNN version used is 3.2. Even with older 2.7.3 version we see this.

Attaching the oneDNN verbose logs here.
[GR_cnnf32_verbose321.txt](https://github.com/oneapi-src/oneDNN/files/12573423/GR_cnnf32_verbose321.txt)
[ICL_f32_verbose321.txt](https://github.com/oneapi-src/oneDNN/files/12573424/ICL_f32_verbose321.txt)
[GR_cnnf32_verbose.txt](https://github.com/oneapi-src/oneDNN/files/12573437/GR_cnnf32_verbose.txt)
[ICL_cnnf32_verbose.txt](https://github.com/oneapi-src/oneDNN/files/12573436/ICL_cnnf32_verbose.txt)

**Comment by abhishek-rn** [comment]
Hi @nSircombe ,
Request your views on this, since now the issue seems to stem from the oneDNN source itself.

Thanks

**Comment by nSircombe** [comment]
The use of the ACL backend, with "fixed format" kernels, means re-orders are required with ACL, to re-order into the correct memory format for ACL's GEMM kernels, which are not needed with the x86 route (see refs. above).
Ultimately, the two backends are different, and are going to produce different VERBOSE logs. I would not expect them to be identical.

Can I ask for a bit of context for your investigation, it might help with our answers to know your motivation & goals, and could help avoid unnecessary work.

**Comment by abhishek-rn** [comment]
Thanks @nSircombe  for your inputs.
We are trying to find differences in frameworks between aarch64 and x64.
So that we can identify scope for improvements in aarch64 compared to a more mature x64 systems.

If ACL backend causes these reorders in large numbers, then that becomes area of improvements.

**Comment by nSircombe** [comment]
Hi @abhishek-rn,

You say "we" - I was wondering who "we" is? I don't see any details in GitHub. It's just that it may be more efficient to discuss some of these nuances, and the bigger performance picture, outside of a GitHub thread.


> If ACL backend causes these reorders in large numbers, then that becomes area of improvements.

yes that's a fair point, but it's important to understand that the re-order happens _somewhere_ however it's engineered on AArch64 with ACL since the GEMM kernels in ACL will need their input re-odered into the format they expect. Without the 'fixed format' kernels, this is still the case, but the re-order happens within ACL and does not show on the oneDNN logs. In adding support for 'fixed format' kernels, this step (re-ordering into the format required by ACL's GEMM kernels) is necessarily exposed in oneDNN. That said, there is room for performance improvements in re-order for sure. And there may be a case for enabling both paths (reorder in oneDNN and reorder within ACL) depending on the use case. But this would need careful consideration as TensorFlow is dependent on the current impl. to enable caching.

**Comment by nSircombe** [comment]
> Here, "we" as in Team at Fujitsu.

Ah, thanks for confirming, I wasn't sure!
